<h1 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #1f2937 45%, #334155 100%);
  padding: 16px 20px;
  border-radius: 14px;
  text-align: center;
  box-shadow: 0 12px 30px rgba(0,0,0,.35);
  font-family: Arial, sans-serif;
  margin: 0;
">
K Vecinos Mas Cercanos - Clasificacion
</h1>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Carga de librerias y del dataset
</h2>

In [17]:
import pandas as pd
import pickle

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [ ]:
url = "https://raw.githubusercontent.com/4GeeksAcademy/naive-bayes-project-tutorial/main/playstore_reviews.csv"
df = pd.read_csv(url)

print("Paso 1 - Dataset cargado")
print(f"Filas y columnas: {df.shape}")
display(df.head())

Paso 1 - Dataset cargado
Filas y columnas: (891, 3)


,package_name,review,polarity
0,com.facebook.katana,privacy at least put some option appear offli...,0
1,com.facebook.katana,"messenger issues ever since the last update, ...",0
2,com.facebook.katana,profile any time my wife or anybody has more ...,0
3,com.facebook.katana,the new features suck for those of us who don...,0
4,com.facebook.katana,forced reload on uploading pic on replying co...,0


<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Seleccion de variables y limpieza del texto
</h2>

In [5]:
df = df[["review", "polarity"]].copy()

df["review"] = df["review"].astype(str).str.strip().str.lower()

print(df.shape)
df.head()

(891, 2)


,review,polarity
0,privacy at least put some option appear offlin...,0
1,"messenger issues ever since the last update, i...",0
2,profile any time my wife or anybody has more t...,0
3,the new features suck for those of us who don'...,0
4,forced reload on uploading pic on replying com...,0


<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
Aqui elimino la columna <b style="color:#D11A2A;">package_name</b> porque no aporta al analisis del sentimiento. Tambien limpio la columna <b style="color:#D11A2A;">review</b> para dejar el texto en un formato mas ordenado antes de convertirlo a numeros.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Separacion de variables y division en train y test
</h2>

In [6]:
X = df["review"]
y = df["polarity"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape, X_test.shape)
print(y_train.shape, y_test.shape)

(712,) (179,)
(712,) (179,)


<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Vectorizacion del texto
</h2>


In [7]:
vectorizer = CountVectorizer(stop_words="english")

X_train_vec = vectorizer.fit_transform(X_train).toarray()
X_test_vec = vectorizer.transform(X_test).toarray()

print(X_train_vec.shape)
print(X_test_vec.shape)

(712, 3310)
(179, 3310)


<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
Aqui convierto el texto en numeros. Cada resena pasa a una fila de valores donde el modelo puede medir que palabras aparecen. El ajuste se hace con <b style="color:#D11A2A;">train</b> y luego esa misma transformacion se aplica a <b style="color:#D11A2A;">test</b>.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Entrenamiento de los modelos Naive Bayes
</h2>

In [8]:
gaussian_nb = GaussianNB()
multinomial_nb = MultinomialNB()
bernoulli_nb = BernoulliNB()

gaussian_nb.fit(X_train_vec, y_train)
multinomial_nb.fit(X_train_vec, y_train)
bernoulli_nb.fit(X_train_vec, y_train)

gaussian_train_pred = gaussian_nb.predict(X_train_vec)
gaussian_test_pred = gaussian_nb.predict(X_test_vec)

multinomial_train_pred = multinomial_nb.predict(X_train_vec)
multinomial_test_pred = multinomial_nb.predict(X_test_vec)

bernoulli_train_pred = bernoulli_nb.predict(X_train_vec)
bernoulli_test_pred = bernoulli_nb.predict(X_test_vec)

<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
Aqui entreno las tres versiones de <b style="color:#D11A2A;">Naive Bayes</b> para compararlas despues con los mismos datos. Asi puedo ver cual se adapta mejor a este problema de clasificacion de texto.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Comparacion de resultados de Naive Bayes
</h2>

In [11]:
naive_bayes_results = pd.DataFrame({
    "model": ["GaussianNB", "MultinomialNB", "BernoulliNB"],
    "accuracy_train": [
        accuracy_score(y_train, gaussian_train_pred),
        accuracy_score(y_train, multinomial_train_pred),
        accuracy_score(y_train, bernoulli_train_pred)
    ],
    "accuracy_test": [
        accuracy_score(y_test, gaussian_test_pred),
        accuracy_score(y_test, multinomial_test_pred),
        accuracy_score(y_test, bernoulli_test_pred)
    ]
})

naive_bayes_results = naive_bayes_results.sort_values(by="accuracy_test", ascending=False).reset_index(drop=True)
naive_bayes_results

,model,accuracy_train,accuracy_test
0,MultinomialNB,0.960674,0.815642
1,GaussianNB,0.985955,0.804469
2,BernoulliNB,0.919944,0.770950


<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
En esta comparacion, el mejor resultado lo dio <b style="color:#D11A2A;">MultinomialNB</b> con un <b style="color:#D11A2A;">accuracy_test de 0.815642</b>. <b style="color:#D11A2A;">GaussianNB</b> quedo un poco por debajo con <b style="color:#D11A2A;">0.804469</b> y <b style="color:#D11A2A;">BernoulliNB</b> fue el mas bajo con <b style="color:#D11A2A;">0.770950</b>. Esto muestra que, para este problema de texto, <b style="color:#D11A2A;">MultinomialNB</b> fue la opcion que mejor generalizo.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Evaluacion del mejor modelo
</h2>

In [12]:
best_model = MultinomialNB()
best_model.fit(X_train_vec, y_train)

y_pred = best_model.predict(X_test_vec)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.84      0.90      0.87       126
           1       0.73      0.60      0.66        53

    accuracy                           0.82       179
   macro avg       0.79      0.75      0.77       179
weighted avg       0.81      0.82      0.81       179

[[114  12]
 [ 21  32]]


<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
El modelo logro un <b style="color:#D11A2A;">accuracy de 0.82</b>, asi que tuvo un rendimiento bueno en general. Detecto mejor las resenas <b style="color:#D11A2A;">negativas</b> que las <b style="color:#D11A2A;">positivas</b>, porque para la clase <b style="color:#D11A2A;">0</b> alcanzo un <b style="color:#D11A2A;">recall de 0.90</b>, mientras que para la clase <b style="color:#D11A2A;">1</b> fue de <b style="color:#D11A2A;">0.60</b>. En la matriz de confusion se ve esto claro: clasifico bien <b style="color:#D11A2A;">114</b> resenas negativas y <b style="color:#D11A2A;">32</b> positivas, pero fallo mas en las positivas, confundiendo <b style="color:#D11A2A;">21</b> de ellas.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Prueba con Random Forest
</h2>

In [13]:
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train_vec, y_train)

rf_train_pred = rf_model.predict(X_train_vec)
rf_test_pred = rf_model.predict(X_test_vec)

print("accuracy_train:", accuracy_score(y_train, rf_train_pred))
print("accuracy_test:", accuracy_score(y_test, rf_test_pred))

accuracy_train: 1.0
accuracy_test: 0.7988826815642458


<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
<b style="color:#D11A2A;">RandomForest</b> logro un <b style="color:#D11A2A;">accuracy_train de 1.0</b>, pero en <b style="color:#D11A2A;">test bajo a 0.798883</b>. Esto muestra un <b style="color:#D11A2A;">sobreajuste claro</b>, porque aprendio demasiado bien los datos de entrenamiento pero no rindio mejor con datos nuevos. Ademas, quedo por debajo de <b style="color:#D11A2A;">MultinomialNB</b>, que habia alcanzado un <b style="color:#D11A2A;">accuracy_test de 0.815642</b>.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Comparacion entre los modelos probados
</h2>

In [14]:
model_comparison = pd.DataFrame({
    "model": ["MultinomialNB", "RandomForest"],
    "accuracy_train": [0.960674, 1.000000],
    "accuracy_test": [0.815642, 0.798883]
})

model_comparison

,model,accuracy_train,accuracy_test
0,MultinomialNB,0.960674,0.815642
1,RandomForest,1.000000,0.798883


<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
Al comparar los dos modelos, <b style="color:#D11A2A;">MultinomialNB</b> sigue siendo la mejor opcion porque logro un <b style="color:#D11A2A;">accuracy_test de 0.815642</b>, mientras que <b style="color:#D11A2A;">RandomForest</b> se quedo en <b style="color:#D11A2A;">0.798883</b>. Aunque RandomForest aprendio perfecto el entrenamiento, en datos nuevos rindio peor.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Prueba con Logistic Regression
</h2>

In [18]:
log_model = LogisticRegression(random_state=42, max_iter=1000)
log_model.fit(X_train_vec, y_train)

log_train_pred = log_model.predict(X_train_vec)
log_test_pred = log_model.predict(X_test_vec)

print("accuracy_train:", accuracy_score(y_train, log_train_pred))
print("accuracy_test:", accuracy_score(y_test, log_test_pred))

accuracy_train: 0.9985955056179775
accuracy_test: 0.8324022346368715


<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
<b style="color:#D11A2A;">LogisticRegression</b> obtuvo el mejor resultado en <b style="color:#D11A2A;">test</b> con un <b style="color:#D11A2A;">accuracy de 0.832402</b>. Supero a <b style="color:#D11A2A;">MultinomialNB</b>, que habia logrado <b style="color:#D11A2A;">0.815642</b>, y tambien a <b style="color:#D11A2A;">RandomForest</b>, que se quedo en <b style="color:#D11A2A;">0.798883</b>. Por tanto, con los resultados obtenidos en este proyecto, <b style="color:#D11A2A;">LogisticRegression</b> fue el modelo que mejor rindio con datos nuevos.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Comparacion final de modelos
</h2>

In [19]:
final_model_comparison = pd.DataFrame({
    "model": ["MultinomialNB", "RandomForest", "LogisticRegression"],
    "accuracy_train": [0.960674, 1.000000, 0.998596],
    "accuracy_test": [0.815642, 0.798883, 0.832402]
}).sort_values(by="accuracy_test", ascending=False).reset_index(drop=True)

final_model_comparison

,model,accuracy_train,accuracy_test
0,LogisticRegression,0.998596,0.832402
1,MultinomialNB,0.960674,0.815642
2,RandomForest,1.000000,0.798883


<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
La comparacion final muestra que <b style="color:#D11A2A;">LogisticRegression</b> fue el mejor modelo, porque logro el mayor <b style="color:#D11A2A;">accuracy_test: 0.832402</b>. Despues quedo <b style="color:#D11A2A;">MultinomialNB</b> con <b style="color:#D11A2A;">0.815642</b> y por ultimo <b style="color:#D11A2A;">RandomForest</b> con <b style="color:#D11A2A;">0.798883</b>.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Guardado del mejor modelo
</h2>

In [20]:
with open("../models/logistic_regression_sentiment_classifier.sav", "wb") as f:
    pickle.dump(log_model, f)

with open("../models/count_vectorizer.sav", "wb") as f:
    pickle.dump(vectorizer, f)

<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
Aqui guardo el mejor modelo, que fue <b style="color:#D11A2A;">LogisticRegression</b>, junto con el <b style="color:#D11A2A;">CountVectorizer</b>. Esto es importante porque el texto nuevo debe pasar por la misma transformacion antes de ser clasificado.
</div>

<h2 style="
  color: #ffffff;
  background: linear-gradient(135deg, #0b0f12 0%, #2b3136 55%, #4a535a 100%);
  padding: 14px 18px;
  border-radius: 12px;
  text-align: center;
  box-shadow: 0 10px 28px rgba(0,0,0,.18);
  border: 1px solid rgba(255,255,255,.10);
  border-top: 5px solid #38bdf8;
  margin: 0;
  font-size: 20px;
">
Conclusion final
</h2>

<div style="
  font-size: 16px;
  line-height: 1.7;
  color: #111827;
  background-color: #ececec;
  border-left: 6px solid #22c55e;
  padding: 16px 18px;
  border-radius: 10px;
  box-shadow: 0 4px 14px rgba(0,0,0,0.08);
  margin-top: 12px;
">
En este proyecto prepare las resenas, convierti el texto en variables numericas y compare varios modelos para clasificar sentimientos. Entre las opciones probadas, <b style="color:#D11A2A;">LogisticRegression</b> fue la que dio el mejor resultado en datos nuevos con un <b style="color:#D11A2A;">accuracy_test de 0.832402</b>. <b style="color:#D11A2A;">MultinomialNB</b> tambien rindio bien con <b style="color:#D11A2A;">0.815642</b>, mientras que <b style="color:#D11A2A;">RandomForest</b> quedo por debajo con <b style="color:#D11A2A;">0.798883</b>. Por tanto, el mejor modelo final de este proyecto fue <b style="color:#D11A2A;">LogisticRegression</b>.
</div>